yjit_codegen.c

```diff
@@ -3384,9 +3385,114 @@ gen_send_iseq(jitstate_t *jit, ctx_t *ctx, const struct rb_callinfo *ci, const r
+            // the caller-specified keyword arguments. If they're not in the
+            // same order as the order specified in the callee declaration, then
+            // we're going to need to generate some code to swap values around
+            // on the stack.
+            ID *caller_kwargs = ALLOCA_N(VALUE, req_key_num);
+            for (int kwarg_idx = 0; kwarg_idx < req_key_num; kwarg_idx++)
+                caller_kwargs[kwarg_idx] = SYM2ID(kw_arg->keywords[kwarg_idx]);
+
+            // First, we're going to be sure that the names of every
+            // callee-specified keyword argument correspond to a name in the
+            // list of caller-specified keyword arguments.
+            for (int callee_idx = 0; callee_idx < req_key_num; callee_idx++) {
+                int caller_idx;
+
+                for (caller_idx = 0; caller_idx < req_key_num; caller_idx++) {
+                    if (callee_kwargs[callee_idx] == caller_kwargs[caller_idx]) {
+                        break;
+                    }
+                }
+
+                // If the keyword was never found, then we know we have a
+                // mismatch in the names of the keyword arguments, so we need to
+                // bail.
+                if (caller_idx == req_key_num) {
+                    GEN_COUNTER_INC(cb, send_iseq_kwargs_mismatch);
+                    return YJIT_CANT_COMPILE;
+                }
+            }
+
+            // Next, we're going to loop through every keyword was that
+            // specified by the caller and make sure that it's in the correct
+            // place. If it's not we're going to swap it around with another one.
+            for (int kwarg_idx = 0; kwarg_idx < req_key_num; kwarg_idx++) {
+                ID callee_kwarg = callee_kwargs[kwarg_idx];
+
+                // If the argument is already in the right order, then we don't
+                // need to generate any code since the expected value is already
+                // in the right place on the stack.
+                if (callee_kwarg == caller_kwargs[kwarg_idx]) continue;
+
+                // In this case the argument is not in the right place, so we
+                // need to find its position where it _should_ be and swap with
+                // that location.
+                for (int swap_idx = kwarg_idx + 1; swap_idx < req_key_num; swap_idx++) {
+                    if (callee_kwarg == caller_kwargs[swap_idx]) {
+                        x86opnd_t swap = ctx_stack_opnd(ctx, argc - 1 - swap_idx);
+                        x86opnd_t kwarg = ctx_stack_opnd(ctx, argc - 1 - kwarg_idx);
+
+                        // First we're going to generate the code that is going
+                        // to perform the actual swapping at runtime.
+                        mov(cb, REG1, swap);
+                        mov(cb, R9, kwarg);
+                        mov(cb, swap, R9);
+                        mov(cb, kwarg, REG1);
```

Comment:
I wonder if we could make this a little more compact by using an `xchg` instruction?
https://www.felixcloutier.com/x86/xchg

Just a remark, I understand if that waits until another PR. Correctness first.

Reply:
https://www.agner.org/optimize/ recommends against xchg (though the trade off might be worth it for instruction size)

Reply:
I tried using the `xchg` but we don't have it working with a memory operand and a register operand at the moment. Let's pair on that tomorrow I'd love to use that instruction.

Reply:
Ah good catch. We probably shouldn't use it because with a memory operand it always has the LOCK prefix. From Intel's manual:

> If a memory operand is referenced, the processor’s locking protocol is automatically implemented for the duration of the exchange operation, regardless of the presence or absence of the LOCK prefix or of the value of the IOPL.

Reply:
Aaah if it does locking then probably best not to use it.